components/butterfly/src/server/incarnation_store.rs

```diff
@@ -0,0 +1,204 @@
+// limitations under the License.
+
+//! Provide the means to persist a Supervisor's own incarnation
+//! number across restarts.
+
+use std::fs::{self, File};
+use std::io::{BufWriter, Read, Write};
+use std::path::PathBuf;
+
+use error::{Error, Result};
+use member::DEFAULT_INCARNATION;
+use std::io;
+use std::num;
+
+/// Provide storage of an incarnation number that can persist across
+/// Supervisor restarts.
+#[derive(Clone, Debug)]
+pub struct IncarnationStore {
+    path: PathBuf,
+}
+
+impl From<PathBuf> for IncarnationStore {
+    fn from(path: PathBuf) -> Self {
+        IncarnationStore { path: path }
+    }
+}
+
+impl IncarnationStore {
+    /// Ensure that the `IncarnationStore` is backed by a suitable
+    /// file on disk. If the file does not already exist, create it
+    /// with an initial incarnation number of 0. If the file does
+    /// exist, an error will be returned if the contents cannot be
+    /// parsed.
+    ///
+    /// Returns the incarnation that is currently stored in the
+    /// backing file (even if we just created it, in which case we
+    /// return `0`).
+    pub fn initialize(&self) -> Result<u64> {
+        if !self.path.exists() {
+            let initial_value = DEFAULT_INCARNATION;
+            self.store(initial_value)?;
+            // TODO (CM): set appropriate file permissions here
+            Ok(initial_value)
+        } else {
+            self.retrieve()
+        }
+    }
+
+    /// Returns the incarnation value found within the file.
+    ///
+    /// Returns an error if the file cannot be read or parsed for any
+    /// reason.
+    pub fn retrieve(&self) -> Result<u64> {
+        File::open(&self.path)
+            .map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))
+            .and_then(|mut file| {
+                let mut incarnation = String::new();
+                file.read_to_string(&mut incarnation)
+                    .map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))
+                    .and_then(|_| {
+                        incarnation.trim().parse().map_err(|e: num::ParseIntError| {
+                            Error::IncarnationParse(self.path.clone(), e)
+                        })
+                    })
+            })
+    }
+
+    /// Store the given `new_incarnation` to disk.
+    pub fn store(&self, new_incarnation: u64) -> Result<()> {
+        let tmp = self.path.with_extension("tmp");
+        let f =
+            File::create(&tmp).map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))?;
+        let mut buf = BufWriter::new(f);
+        buf.write(new_incarnation.to_string().as_bytes())
+            .map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))?;
+        buf.flush()
+            .map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))?;
+        fs::rename(&tmp, &self.path)
+            .map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))?;
```

Comment:
I would love to. All the errors I'm converting currently require the path to be present as additional context (which is really handy if a user trying to debug why their Supervisor isn't starting because their incarnation file is somehow corrupted). Unfortunately, the `From::from` trait function only takes a single argument (the original error), so I can't create an implementation that'll add in extra information that is known only at runtime. Since the `From` trait is what the `?` operator uses to do the mapping of error types, I think we're stuck with `map_err` and friends (unless I'm blanking on something obvious from having stared at this too long.)

I could define an inner function that does this mapping, which would reduce some of the verbosity here, but I think we're still stuck in same general situation. 

We could remove the extra context, but I feel like that's got some real user value that I wouldn't want to get rid of to save a few lines of code here.

Reply:
To be concrete, it'd look something like this:

```rust
    pub fn store(&self, new_incarnation: Incarnation) -> Result<()> {
        let into_err = |e: io::Error| Error::IncarnationIO(self.path.clone(), e);

        let tmp = self.path.with_extension("tmp");
        let f = File::create(&tmp).map_err(into_err)?;
        let mut buf = BufWriter::new(f);
        buf.write(new_incarnation.to_string().as_bytes())
            .map_err(into_err)?;
        buf.flush().map_err(into_err)?;
        fs::rename(&tmp, &self.path).map_err(into_err)?;
        Ok(())
    }
```

For me, that's six of one, half-a-dozen of the other; I don't have a strong opinion either way.

Reply:
It's not a big deal. I'll poke around a bit and see if I can realize what I was thinking or if it helps me to understand why that won't work.

Reply:
Removing the `and_then`s is totally doable, though, and looks much better. Not quite sure what I was doing when I wrote that 😄

Reply:
For `store`, it seems like the last line could be
```rust
        buf.write(new_incarnation.to_string().as_bytes())
            .and_then(|_| buf.flush().and_then(|_| fs::rename(&tmp, &self.path)))
            .map_err(|e: io::Error| Error::IncarnationIO(self.path.clone(), e))
```
no?

Maybe it's a little too weird. It does eliminate the redundant error handling, though.